File: ruby/53.maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
  current_sum = ans = nums[0]
  nums[1..].each do |num|
    current_sum = [current_sum + num, num].max
    ans = [ans, current_sum].max
  end

  ans
end

File: rust/src/53.maximum-subarray.rs
use crate::Solution;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut current_sum = nums[0];
        let mut ans = nums[0];

        for &num in nums.iter().skip(1) {
            current_sum = num.max(current_sum + num);
            ans = ans.max(current_sum);
        }

        ans
    }
}
